src/migrate/Migrator.js

```diff
@@ -419,15 +440,42 @@ export default class Migrator {
   }
 
   setConfig(config) {
-    return assign({}, CONFIG_DEFAULT, this.config || {}, config);
+    const newConfig = assign({}, CONFIG_DEFAULT, this.config || {}, config);
+
+    // If migrationSource is not specified, but fs related config has been specified
+    // recreate the FsMigration source so the new filesystem related settings
+    // are used
+    if (
+      config &&
+      !config.migrationSource &&
+      (config.directory ||
+        config.sortDirsSeparately !== undefined ||
+        config.loadExtensions)
+    ) {
+      newConfig.migrationSource = new FsMigrations(
+        newConfig.directory,
+        newConfig.sortDirsSeparately,
+        newConfig.loadExtensions
+      );
+    }
+
+    // Ensure a migrationSource is set
+    if (!newConfig.migrationSource) {
```

Comment:
This code looks quite redundant. What's the added benefit of Calling F1 with A, B, C if A || B || C set, or calling F1 with A, B if neither are set, if these are full equivalents, as you would simply be passing undefineds that will get resolved to default value in either case anyway?

Reply:
This method indeed seems unnecessarily complicated... couldn't we just always do

```js
if (!newConfig.migrationSource) {
    newConfig.migrationSource = new FsMigrations(
        newConfig.directory,
        newConfig.sortDirsSeparately,
        newConfig.loadExtensions
    );
}
```

?

Reply:
Ah yeah, I think now I get it... anyways, maybe it would be just better to do first

```js
if (config && (config.directory ||
        config.sortDirsSeparately !== undefined ||
        config.loadExtensions) {
   newConfig.migrationSource = null;
}
```

Which would signal the intention better that if user has passed file migrator configuration attributes to parameter, then we need to override the migration source.

Reply:
Ah that is actually quite a nice way of doing it. Invalidate the current if one of the config settings change, then set if not set. Will update!